### PR TITLE
release-tools: add debian-package-builder

### PR DESCRIPTION
**Pull request by zyga**
Building Debian packages in a clean, efficient and repeatable way is not
very difficult but requires some know-how that is easier to share in the
form of a small executable document.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: release-tools/debian-package-builder [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+.PHONY: create-debian-10
+create-debian-10:
+	multipass launch -n debian-10 -c 8 -m 4G https://cdimage.debian.org/mirror/cdimage/openstack/testing/debian-testing-openstack-amd64.qcow2
+
+.PHONY: prepare-debian-9 prepare-debian-10
+prepare-debian-9 prepare-debian-10: prepare-%:
+	$(remote) sudo mkdir -p /etc/apt/apt.conf.d/
+	$(remote) sudo sh -c 'echo "Acquire::HTTP::Proxy \"$(proxy_url)\";" > /etc/apt/apt.conf.d/00proxy'
+	-$(remote) sudo apt-get update
```
Comment (zyga) [inline release-tools/debian-package-builder]: NOTE: this fails on sid, because of missing Release file in one of the repositories.

**Comment by bboozzoo** [comment]
```
release-tools/debian-package-builder:3:40: "occuring" is a misspelling of "occurring"
release-tools/debian-package-builder:8:43: "conveniece" is a misspelling of "convenience"
```

File: release-tools/debian-package-builder [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# This program allows one to, relatively easily, iterate on debian packaging in
+# a way that is extremely close to that occurring in the real Debian archive.
+#
+# The key aspect is that packaging is done on a machine running Debian (either
+# 9 or 10), with the real Debian kernel, with real build-packages (that is,
+# devoid of any patches potentially present only in Debian), with correct
+# pristine build environment, unspoiled by convenience packages often present on
+# workstation environments. 
+#
+# The systme is "portable", being able to execute on anything capable of
```
Comment (stolowski) [inline release-tools/debian-package-builder]:
```suggestion
# The system is "portable", being able to execute on anything capable of
```

File: release-tools/debian-package-builder [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	@echo "  destroy-NAME: destroy a multipass virtual machine"
+	@echo "  build-on-NAME: copy a source package and build it on a virtual machine"
+	@echo
+	@echo "Available names are: debian-9 debian-10"
+	@echo "NOTE: You must provide the source package yourself"
+	@echo "Please read the script for additional instructions."
+
+.PHONY: create-debian-9
+create-debian-9:
+	multipass launch -n debian-9 -c 8 -m 4G https://cdimage.debian.org/mirror/cdimage/openstack/current/debian-9.6.4-20190118-openstack-amd64.qcow2
```
Comment (stolowski) [inline release-tools/debian-package-builder]: I'd extract this URL (and the one below) into variables defined close the top of file for easy edits (maybe even allow them to be overriden by commandline/env - it's semi-automatic with makefiles AFAIR, if declared properly).

File: release-tools/debian-package-builder [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+# - prepare the source package:
+#   On appropriate branch of the git tree compatible with
+#   building the debian package, e.g. https://salsa.debian.org/zyga-guest/snapd
+#   on the "debian" branch, issue the command
+#   $ apt-get build-dep ./
+#   $ dpkg-buildpackage -S
+#   TIP: you can iterate on the packing and repeat this step!
+#   If you need to make patches to upstream snapd parts use the "debian-patches"
+#   branch from the same repository, where history is broken down. Then use
+#   $ git format-patches 2.37..HEAD
```
Comment (chipaca) [inline release-tools/debian-package-builder]: always 2.37?

File: release-tools/debian-package-builder [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/make -f
+# This program allows one to, relatively easily, iterate on debian packaging in
+# a way that is extremely close to that occurring in the real Debian archive.
+#
+# The key aspect is that packaging is done on a machine running Debian (either
+# 9 or 10), with the real Debian kernel, with real build-packages (that is,
+# devoid of any patches potentially present only in Debian), with correct
```
Comment (chipaca) [inline release-tools/debian-package-builder]: *devoid* of any patches?

Comment (zyga) [reply release-tools/debian-package-builder]: I meant Ubuntu :)

**Comment by zyga** [comment]
Having discussed this with @mvo5 today we agreed to move this to a debug section of the spread test that uses `sbuild` to compile the package. This will not prevent the code from bit-rotting but at least will put the relevant "this is how to iterate" information right next to the failing test.

**Comment by mvo5** [comment]
@zyga We can keep it in the current place and add a note in the debug section that this should be used to debug sbuild failures.

File: release-tools/debian-package-builder (hunk 3 above)
Comment (zyga) [reply release-tools/debian-package-builder]: Done now.

File: release-tools/debian-package-builder (hunk 4 above)
Comment (zyga) [reply release-tools/debian-package-builder]: Clarified now.

File: release-tools/debian-package-builder [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+# devoid of any patches potentially present only in Ubuntu), with correct
+# pristine build environment, unspoiled by convenience packages often present on
+# workstation environments. 
+#
+# The system is "portable", being able to execute on anything capable of
+# running Multipass and apt-cacher-ng. The workflow is as follows:
+#
+# - prepare host environment:
+#   $ snap install multipass --classic --edge
+#   $ apt install apt-cacher-ng
```
Comment (bboozzoo) [inline release-tools/debian-package-builder]: It'd be nice to have a switch/option to not use apt-cacher-ng if it's not in the host distro repositories (as it isn't in Arch). Not a blocker though.

Comment (zyga) [reply release-tools/debian-package-builder]: Without this iteration is extremely painful. I may explore a snap option.

File: release-tools/debian-package-builder [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+#   In case the build fails an interactive shell will be spawned by sbuild, still
+#   running inside the virtual machine. The package is unpacked twice, you
+#   really only care about the files in the _build directory. You can set
+#   GOPATH manually to
+#   GOPATH=/path/to/directory/called/_build:/usr/share/gocode and run tests and
+#   iterate as usual. The source tree alongside the _build directory is useful
+#   for comparison and for generating patches..
+
+# There must be an apt-cacher-ng installation on the host.
+proxy_url=http://$(shell ip --json addr list scope global up | jq --raw-output '.[0].addr_info[0].local'):3142
```
[the remark this answers is not in the document]
Comment (zyga) [reply release-tools/debian-package-builder]: Yeah, any suggestions?

[the remark this answers is not in the document]
Comment (bboozzoo) [reply release-tools/debian-package-builder]: Looks like multipass is use `mpqemubr0` for the bridge interface name. I'd try something like this:
```
ip --json addr show dev mpqemubr0 | \
    jq --raw-output '.[0].addr_info | .[] | select(.family=="inet") | .local'
```

[the remark this answers is not in the document]
Comment (zyga) [reply release-tools/debian-package-builder]: I rewrote the expression based on your feedback.

**Pull request merged**